File: data/EXP01/python/2069/2069.output_3.py
# @lc app=leetcode id=2069 lang=python3
class Robot:

    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        self.x = 0
        self.y = 0
        self.direction = "East"

    def step(self, num: int) -> None:
        for i in range(num):
            if self.direction == "North":
                if self.y + 1 >= self.height:
                    self.direction = "West"
                else:
                    self.y += 1
            elif self.direction == "East":
                if self.x + 1 >= self.width:
                    self.direction = "North"
                else:
                    self.x += 1
            elif self.direction == "South":
                if self.y - 1 < 0:
                    self.direction = "East"
                else:
                    self.y -= 1
            elif self.direction == "West":
                if self.x - 1 < 0:
                    self.direction = "South"
                else:
                    self.x -= 1

    def getPos(self) -> List[int]:
        return [self.x, self.y]

    def getDir(self) -> str:
        return self.direction

File: data/EXP01/golang_no_system_var/2069/2069.output_3.go
type Robot struct {
    width  int
    height int
    x      int
    y      int
    dir    string
}

func Constructor(width int, height int) Robot {
    return Robot{
        width:  width,
        height: height,
        x:      0,
        y:      0,
        dir:    "East",
    }
}

func (r *Robot) Step(num int) {
    for i := 0; i < num; i++ {
        nx, ny := r.getNextPosition()
        if r.isOutOfBounds(nx, ny) {
            r.turnCounterclockwise()
            nx, ny = r.getNextPosition()
        }
        r.x = nx
        r.y = ny
    }
}

func (r *Robot) GetPos() []int {
    return []int{r.x, r.y}
}

func (r *Robot) GetDir() string {
    return r.dir
}

func (r *Robot) getNextPosition() (int, int) {
    switch r.dir {
    case "North":
        return r.x, r.y + 1
    case "East":
        return r.x + 1, r.y
    case "South":
        return r.x, r.y - 1
    case "West":
        return r.x - 1, r.y
    }
    return r.x, r.y
}

func (r *Robot) isOutOfBounds(x, y int) bool {
    return x < 0 || x >= r.width || y < 0 || y >= r.height
}

func (r *Robot) turnCounterclockwise() {
    switch r.dir {
    case "North":
        r.dir = "West"
    case "East":
        r.dir = "North"
    case "South":
        r.dir = "East"
    case "West":
        r.dir = "South"
    }
}